### ISSUE TITLE
Vips::Error: No known saver for '/tmp/image_processing20200724-11554-lsl5p1.jfif'

### ISSUE THREAD

**Issue by johan-smits** [rails/rails#39921]
### Steps to reproduce

```ruby
class SomeModel < ApplicationRecord
  has_one_attached :file
end

model = SomeModel.create!
model.file.attach(io: File.open('image.jfif'), filename: 'image.jfif', content_type: 'image/jpeg')
model.file.representation({resize_to_fit: [1920, 1920]}).processed
```

### Expected behavior
Generate a representation

### Actual behavior
```
Error:
CollectionTest#test_jfif_file_format:
Vips::Error: No known saver for '/tmp/image_processing20200724-11554-lsl5p1.jfif'.
```

### System configuration
**Rails version**: 6.0.3.1

**Ruby version**: 2.5.1p57

Used image: [image.zip](https://github.com/rails/rails/files/4972128/image.zip)

**Comment by victorperez** [comment]
@johan-smits this looks like related with `ruby-vips` gem not with `rails` itself, here you have more information: https://github.com/libvips/ruby-vips/issues/133

**Comment by johan-smits** [comment]
@victorperez if this is a known thing should rails preview then nog select the desired extension itself and not assume the extension of the source file?

Because the source file could be: image.foo (content type: image/jpg) but you want a hash-image.jpg? So would it be sensible to use the last part of content type as the extension for vips?

**Comment by victorperez** [comment]
@johan-smits I didn't fully review the issue and the current active storage implementation but what you mention should be possible, although using the last part of the content type is not possible in all the scenarios since it doesn't necessarily match with the correct extension. 

I will investigate it deeper at some point to see if a fix in the rails code is possible.

**Comment by johan-smits** [comment]
@victorperez thanks, or in case of a image it could do something special. Because the mime types where the variant is working for is known and declared. So it could best effort fix it.

**Comment by phylor** [comment]
For all Googlers: upgrading to Rails 6.1 solves this issue.

**Comment by johan-smits** [comment]
@phylor it is not really fixed in `6.1`. When you use vips you can get the error:

```
A Vips::Error occurred in background at 2023-09-09 02:33:04 UTC :

  VipsForeignSave: "/tmp/image_processing20230909-39-s4f4a0.jfif" is not a known file format
VipsForeignSave: "/tmp/image_processing20230909-39-s4f4a0.jfif" is not a known file format

  /var/app/current/vendor/bundle/ruby/3.0.0/gems/ruby-vips-2.1.4/lib/vips/image.rb:590:in `write_to_file'
```
This is because vips does not know the extension. When you call a representation you can ask for a specific format like this: `item.file.representation({ resize_to_fit: [1920, 1920], format: :jpeg })`.

**Comment by KevinBerthier** [comment]
Have the same issue as @johan-smits  for JFIF. Any way to configure Active Storage to convert JFIF to JPEG at a global level ?

**Comment by johan-smits** [comment]
@KevinBerthier as fas as I know not. Here is the commit that introduced the feature: https://github.com/rails/rails/commit/a185ff86c0b46c7a643b9c8a9ffd4d973c1ab9dd

**Comment by johan-smits** [comment]
@pranavrajs you are correct.

@georgeclaghorn to me the issue should be reopened because it is not fixed.

**Comment by skipkayhil** [comment]
Can you open a new issue with a full reproduction?

**Comment by northeastprince** [comment]
I'm having the [same `Vips::Error`s](https://github.com/rails/rails/issues/39921#issuecomment-1715391529) on Rails edge - content type is being detected as JPEG but the JFIF extension is causing it.

**Comment by brendon** [comment]
For some reason this has only just started cropping up for us with images already saved a long time ago. Can't really put my finger on the reason but we did upgrade Rails from 6.1.7.3 to 6.1.7.6.

**Comment by brendon** [comment]
Renaming the file name in `active_storage_blobs` from `whatever.jfif` to `whatever.jpeg` fixes the error. Still a mystery as to why this suddenly became a thing though.

**Comment by northeastprince** [comment]
Yeah, I'm getting the same error on Rails edge:

`VipsForeignSave: "/tmp/image_processing20231028-48-tqj9zq.jfif" is not a known file format`

Remaining the file extension to jpeg fixes it.

@zzak can this be reopened?

**Comment by zzak** [comment]
I have a [failing test](https://buildkite.com/rails/rails/builds/101408#018b83ab-6ccc-4d5f-8e63-827bef174b21/1090-1147) but we need to think about how to fix it especially #39984 doesn't _feel_ like the correct approach

**Comment by Augustin-Grenne** [comment]
I'm not quite sure this is the right place for it, apologies if it's not, but I seem to have a similar issue.
We've recently switched our Rails app from using ImageMagick to Vips and ever since our Slack error channel is flooded with similar errors, e.g. `VipsForeignSave: "/tmp/image_processing20231110-28-d3zg11.jfif" is not a known file format`. 
It's not just this one file. Most images in our app are uploads from users coming from various points in our application. I've tried to track down a few of them, but by the time I've found which image seems to be responsible everything appears to be resolved, the image is attached and renders without any further errors.

@zzak maybe this is another cause for reopening this issue?

I'd be happy to provide more information about our setup etc. to anyone who can help me understand more about what's going on, and maybe put a stop to this bombardment of errors.

**Comment by chaadow** [comment]
Hi @Augustin-Grenne @johan-smits @northeastprince  and everyone involved,

I believe I've fixed the issue in #50175 Could you please confirm that my fix works for you? Thanks!

**Comment by TiuTalk** [comment]
I had a similar issue and, while this is not a permanent fix, it does help fixing the existing data so VIPs can handle the image processing:

```ruby
ActiveStorage::Blob.where("filename ILIKE ?", "%.jfif").find_each { |blob| blob.update!(filename: blob.filename.to_s.gsub(/\.jfif/i, '.jpeg')) }
```

**Comment by Augustin-Grenne** [comment]
@chaadow Sorry for the wait. I finally had some time to do some testing. I wasn't able to test your change with our full app, but I checked your code in a GitHub codespace and made a small app mimicking the most relevant aspects of our app. Using a jfif file that caused the error in our app (seen both on production and local development) I could confirm the error doesn't appear and the image is displayed as expected. In short, I can't say with 100% certainty that it will, but it strongly looks like this would resolve the issue for us. Thanks for your work! 🙏 I hope your fix will get merged soon.

**Comment by casperbrike** [comment]
Recently we stumbled upon similar this issue and discovered that the root cause lies inside libvips. According to backtrace, Active Storage tries to call `Vips::Image#write_to_file` method, this method in turn calls `vips_foreign_find_save` function from libvips and the function raises `VipsForeignSave` error for JFIF images. Interestingly, libvips does support JFIF format, for example it's possible to convert from JPG to JFIF and vice versa using CLI:
```
vips jpegsave img.jfif converted.jpg
vips jpegsave converted.jpg converted.jfif
```

So I'd say it's not really a issue with Active Storage and it'd be better to address it at libvips level. I've already created a PR at their repo: https://github.com/libvips/libvips/issues/3775

**Comment by zzak** [comment]
@casperbrike Thank you! I was just talking with @chaadow last night about reporting this upstream before trying to make any changes to Rails, so you saved us a ton. 🙏

**Comment by chaadow** [comment]
@zzak and everyone involved https://github.com/libvips/libvips/issues/3775 . It will be released in `8.15.1` ( not released yet)

**Comment by zzak** [comment]
Closing this since the bug has been fixed and released upstream in libvips [8.15.1](https://github.com/libvips/libvips/releases/tag/v8.15.1).

For those who are unable to upgrade their libvips, there are some workarounds in this thread, or you could try monkey-patching Active Storage similar to: https://github.com/rails/rails/commit/6f0276115886cf1529f8c7ae67cf30de567a4c5c

```diff
diff --git a/activestorage/lib/active_storage/transformers/image_processing_transformer.rb b/activestorage/lib/active_storage/transformers/image_processing_transformer.rb
index fe33cc638c2e8..dd14d124fc998 100644
--- a/activestorage/lib/active_storage/transformers/image_processing_transformer.rb
+++ b/activestorage/lib/active_storage/transformers/image_processing_transformer.rb
@@ -20,7 +20,7 @@ def process(file, format:)
           processor.
             source(file).
             loader(page: 0).
-            convert(format).
+            convert(reformat(format)).
             apply(operations).
             call
         end
@@ -105,6 +105,12 @@ def validate_arg_hash(argument)
             end
           end
         end
+
+        def reformat(format)
+          return format unless ActiveStorage.variant_processor == :vips
+          
+          { "jfif" => "jpg" }.fetch(format, format)
+        end
     end
   end
 end
 ```
